[DRAFT] Feature/hadisd dataset integration

millerjoel opened the pull request:
Added a HadISD data accessor along with some custom pipeline steps and a notebook to demonstrate the use of the accessor

millerjoel committed:
First implementation of HadISD integration using ERA5 as template (untested)

millerjoel committed:
First implementation of HadISD integration using ERA5 as template (untested)

millerjoel committed:
Updata HadISDIndex class: Can now successfully load Hadisd data from within PET

millerjoel committed:
Create notebook for documenting HadISD dataset integration

millerjoel committed:
Update HadISDIndex class so variables can be selected using transforms

millerjoel committed:
Add small explainer comments

millerjoel committed:
Add transform for selecting variables when loading datasets

millerjoel committed:
Merge branch 'ACCESS-Community-Hub:develop' into feature/hadisd-dataset-integration

millerjoel committed:
Update class to support selecting variables

millerjoel committed:
Add SetMissingToNaN transform class

millerjoel committed:
Update HadISDIndex Class to make use of new SetMissingToNaN transform

millerjoel committed:
Update HadISD note book to reflect addition of new transforms

millerjoel committed:
Add new test for SetMissingToNaN transform

millerjoel committed:
Merge branch 'ACCESS-Community-Hub:develop' into feature/hadisd-dataset-integration

millerjoel committed:
Create add_flagged_obs class in values.py module

millerjoel committed:
Update Hadisd notebook

millerjoel committed:
Add ReIndexTime transform to coordinates.py module

millerjoel committed:
Merge branch 'ACCESS-Community-Hub:develop' into feature/hadisd-dataset-integration

millerjoel committed:
Minor Hadisd changes

millerjoel committed:
Merge branch 'ACCESS-Community-Hub:develop' into feature/hadisd-dataset-integration

millerjoel committed:
Merge branch 'ACCESS-Community-Hub:develop' into feature/hadisd-dataset-integration

millerjoel committed:
Merge branch 'ACCESS-Community-Hub:develop' into feature/hadisd-dataset-integration

millerjoel committed:
Update init to reflect correct archive

millerjoel committed:
Update load and filesystem methods to work with zarr

millerjoel committed:
Update notebook to support loading multiple stations and merging into one dataset

millerjoel committed:
Add support for natively loading multifle zarr directories

millerjoel committed:
Update to test with netcdf

millerjoel committed:
single station testing

millerjoel committed:
Merge branch 'ACCESS-Community-Hub:develop' into feature/hadisd-dataset-integration

millerjoel committed:
Save development changes

millerjoel committed:
Save development

millerjoel committed:
Add rough FeatureTargetSplit operation to notebook. Add to operations later

millerjoel committed:
Merge branch 'ACCESS-Community-Hub:develop' into feature/hadisd-dataset-integration

millerjoel committed:
Refine class and make more readable

millerjoel committed:
change path

millerjoel committed:
revert changes

millerjoel committed:
More development and testing with xgboost and pipeline operations

millerjoel committed:
Add notebook for converting netcdf files to zarr

millerjoel committed:
Merge branch 'ACCESS-Community-Hub:develop' into feature/hadisd-dataset-integration

millerjoel committed:
Updated analysis of qc flags

millerjoel committed:
Add XGBoost training plus data manipulation steps

tennlee commented on packages/data/src/pyearthtools/data/transforms/values.py:
```diff
@@ -66,6 +67,74 @@ def apply(self, dataset: xr.Dataset) -> xr.Dataset:
         return encod(dataset)
+    
+
+class SetMissingToNaN(Transform):
```
Looks good, please add it to the API docs, and the API index table.

tennlee commented on packages/data/src/pyearthtools/data/transforms/values.py:
```diff
@@ -66,6 +67,74 @@ def apply(self, dataset: xr.Dataset) -> xr.Dataset:
+class SetMissingToNaN(Transform):
+    """
+    Transform to replace specified missing values with NaN for given variables.
+
```
Please give a short example in the docstring

tennlee replied:
Looks good overall, please add to the docs as well

tennlee commented on packages/tutorial/src/pyearthtools/tutorial/HadISD_to_zarr.ipynb:
This should be in a notebooks folder, not in a src folder. Or, turn it into a Python script if you want it here.

tennlee commented on packages/data/src/pyearthtools/data/transforms/coordinates.py:
```diff
@@ -206,6 +207,35 @@ def apply(self, dataset: xr.Dataset):
 
+class ReIndexTime(Transform):
+    """Reindex the time coordinate to a complete hourly time series."""
+# TODO look into using Petdt for this
```
Yes, ideally this would support Petdt. I'm a bit unclear about the implications of changing the start and end time here. What happens if the start time isn't the start of the time-series on disk? What happens if the duration on disk isn't a clean multiple of the frequency? What happens if the user doesn't actually know the time range on disk but just wants to set the frequency? What if the start time on disk isn't aligned to the right fenceposts? e.g. if the first time step is 26 minutes into the hour, and someone reindexes to hourly on the 26th minute, is that what's wanted? Maybe just explain that this is using the pandas date_range and default reindexing logic to perform the function, then users can go look over there for details.

millerjoel replied:
That's a good point! In the end I didn't actually use this for anything, but it's a feature that probably still worth thinking about for the future.

tennlee replied:
Feel free to remove it from the pull request if you don't have an immediate use case

tennlee commented on packages/data/src/pyearthtools/data/transforms/values.py:
```diff
@@ -66,6 +67,74 @@ def apply(self, dataset: xr.Dataset) -> xr.Dataset:
+        return dataset
+
+
+class AddFlaggedObs(Transform):
```
Do you think this should go into something like data.transforms.stationdata instead of .values, since this flagging of obs might not apply to gridded or other data sets? Just a thought.

millerjoel replied:
I was initially unsure where to put this, but included it in the values module since it alters the values of the data. But I guess it is the sort of thing that is only applied to station data. Maybe we can talk about this soon.

millerjoel committed:
Move HadISD to Zarr to notebooks

millerjoel committed:
HadISD QC exploration notebook

nikeethr commented on packages/data/src/pyearthtools/data/transforms/values.py:
```diff
@@ -66,6 +67,74 @@ def apply(self, dataset: xr.Dataset) -> xr.Dataset:
+    def apply(self, dataset: xr.Dataset) -> xr.Dataset:
+        for var_name, miss_val in self.varname_val_map.items():
+            if var_name in dataset:
+                dataset[var_name] = dataset[var_name].where(dataset[var_name] != miss_val, np.nan)
```
Ideally filling `nans` is done on opening datasets using the `_FillValue` item or using a pre-processer to specify it - this will reduce the need for having an additional search using `where`. `where` is quite slow with lazy datasets, because it may end up loading more than it needs to.

See `mask_and_scale` and `preprocess` options in `open_dataset` and the like. If this pattern is common in other parts of PET, its fine for consistency sake - noting its not the most portable way of dealing with this - as there are CF conventions that already do this.

_This is more for lazy datasets (i.e. not eagerly using `.load()` or `.compute()` into memory - for those cases it doesn't really matter)._

nikeethr commented on packages/tutorial/src/pyearthtools/tutorial/HadisdDataClass.py:
```diff
@@ -0,0 +1,298 @@
+        base_transform += SetMissingToNaN(varname_val_map)
+
+        super().__init__(
+            transforms=base_transform + (transforms or TransformCollection()),
```
Use explicit `if transforms is None: ... else: ...` instead of Falsey. If we're expecting `None` to be the default we should compare against `None`, a lot of things can be False not just `None`. Since we're not statically typed - "Explicit is better than implicit".

millerjoel committed:
remove notebook outputs

millerjoel committed:
Add pipeline branches, simplify notebook

millerjoel committed:
Add HadISD config notebook for custom pipelinesteps and dictionaries

millerjoel committed:
Remove TODOs

millerjoel committed:
Merge branch 'ACCESS-Community-Hub:develop' into feature/hadisd-dataset-integration

millerjoel committed:
Remove TODOs

millerjoel committed:
add train test split pipeline step

millerjoel committed:
Include train test split in pipeline

millerjoel committed:
Feed output of new pipeline (train/test) to xgboost

millerjoel committed:
Update get all stations method for more universal suport

millerjoel committed:
Update get all stations method for more universal suport

millerjoel committed:
Define user agnostic paths

millerjoel committed:
Add download notebook and rename other HadISD notebooks

millerjoel committed:
Clear notebook outputs

millerjoel committed:
Update notebooks

millerjoel committed:
Merge branch 'ACCESS-Community-Hub:develop' into feature/hadisd-dataset-integration

millerjoel committed:
Add data config notebook

millerjoel committed:
Update notebooks to use config notebook

millerjoel committed:
Update init so hadis works out the box

Commit:
Rename Config notebooks

Commit:
Clear notebook outputs and update Data_config

millerjoel committed:
Remove redundent cells from notebooks

stevehadd commented:
This is looking good @millerjoel . Some general thoughts from running through the notebooks.

* Some of the cells in the notebooks have quite a lot in each cell. I usually find it more readable for each cell to be focused on one thing. For example I'd suggest not having import statements in the same cell as the code that then executes functions from the imported modules. Also generally helpful from a developer point of view, as you often want to rerun something, and haviung smaller cells makes it easier to rerun only the bit you are interested in.
* Although the data source divides the data into 4 files, thats not a fundamental division, just a convenient one so the individual files for download are not too big. I'd suggest that once you provcessed into zarr, there's not reason to maintain that division, so maybe consider having a single "HadISD" zarr at the top level of the dataset (i.e. `~/HadISD/zarr/`) which better maps to to it conceptually being a single dataset, rather than 4 separate zarr archives. That way you should also avoid the need for the `load_combined_dataset` function, as you will just point at the single zarr archive directory, and pyearthtools/xarray will take care of the rest
* I think you've told me this already, but what was the reasoning for defining the functions in the notebook rather than in the core pyearthtools python libraries? It seems like these would be useful functions to be able to reuse, and it would be easier to do so if they were in the python library. @tennlee  perhaps you could comment architecturally? I'm not saying one way or the other is right, just trying to understand the current approach was chosen to inform my own future contributions.
* I'm sure its already on your todo list, but make sure you delete all the commented out code before you merge this PR.
* in the xgboost notebook, (#3) where you specify X_train, y_train etc., so you could rather combine this with the previous cell, so rather have `X_train, y_train), (X_test, y_test) = data_prep_pipe["1969-01-01T00"]`
* Before merging you should also rerun the notebooks in the order of the cells, so that the committed version has example output, so that people can look at the output on github, so they know what to expect when they rerun the notebook.
* maybe something for a follow up, but as part of the purpose of demonstrating the dataset, it might be nice to have a more general data exploration notebook, as one would usually do some data exploration before training an ML model, so as to show people what the dataset contains. In this case we were interested from a data QC perspective, but the idea is this dataset is more broadly applicable, so it would be good to show that in a notebook.

those are just my random thoughts, don't feel you have to act on all or any of them.

millerjoel committed:
Notebooks ready for merge

millerjoel commented:
Thanks @stevehadd! I think I've addressed most of your comments, but will leave some things for another PR. The reason for not having some of the pipeline steps as part of the core library was to demonstrate a way that custom steps can be easily added to a pipeline. In the future I will likely make a separate PR to have some of these steps exist as part of PET.

millerjoel committed:
Update notebook 3

millerjoel committed:
Improve xarray/dask handling based on review comments

millerjoel committed:
Use explicit instead of falsey

tennlee commented:
I think we can merge it as-is but there are some improvements I'd suggest making. 

- [ ] Firstly, in the downloading, it should be possible to do this without manually selecting each time-range one by one, and just select all data. 
- [ ] Secondly, I think there is a confusion in the downloader about the zip filename and the tar filename. I think the incremental downloader is checking the tar filename when it should be checking the zip filename
- [ ] Thirdly, in the converter, it currently raises a lot of warning messages, which are distracting given they are raised for every small netcdf file - I would suggest capturing them and dropping all except the first one. 
- [ ] Consider adding a progress bar to the converter, or a comment on how long it's expected to take
- [ ] I tried this on the entire data set. There are something like 6500 individual station files as far as I can tell, and I think it's doing something like 100-200 per minute, so it's going to take 30-60 minutes I guess. It's also maxing out around 112% CPU, so it's not very multithreaded. I think the notebook needs to explain to people what's involved, and maybe the progress bar is warranted. Alternatively, put a stronger recommendation to start with a single file for experimentation to avoid people trying to do too much at once.
- [ ] I did my initial download with wget, which resulted in some slight differences. In the end I just unzipped and untarred the files at the command-line as a result. That's okay because I know I wasn't following the notebook, but I thought I'd mention it. I also ended up with some corrupted gz files, and I'm not sure which. I wonder how much space the whole dataset takes up as a parquet file, which might be much more efficient than gz in this case. I might try a test later, but it could be that making a .pq distribution file of the dataset and hosting it would lower the barrier significantly for users doing the initial setup.

It's going through the conversion now, I'll add further comments post-conversion once that's done.

tennlee commented on packages/tutorial/src/pyearthtools/tutorial/HadisdDataClass.py:
```diff
@@ -0,0 +1,334 @@
+        }
+    def __init__(
+        self,
+        station: str | list[str] | None = None,  # Allow single station, multiple stations, or None
```
It should be possible to override the base directory for users who have put the data in a different location outside of their home directory

tennlee commented on packages/tutorial/src/pyearthtools/tutorial/HadisdDataClass.py:
```diff
@@ -0,0 +1,334 @@
+        self,
+        station: str | list[str] | None = None,  # Allow single station, multiple stations, or None
+        variables: list[str] | str | None = None, 
+        *,
```
The optional arguments should all be after the "*" in the constructor (probably)

tennlee commented on packages/tutorial/src/pyearthtools/tutorial/HadisdDataClass.py:
```diff
@@ -0,0 +1,334 @@
+    #     Returns:
+    #         list[str]: A list of all station IDs.
+    #     """
+    #     root_directory = Path(root_directory)
```
Please strip out the commented code if it's not a part of the PR

tennlee commented on packages/tutorial/src/pyearthtools/tutorial/HadisdDataClass.py:
```diff
@@ -0,0 +1,334 @@
+                raise DataNotFoundError(f"Root directory does not exist: {root_directory}")
+            station_ids = []
+            for file in cached_iterdir(root_directory):
+                if file.suffix == ".nc":
```
This seems to be getting the station ids from the netcdf files, not the zarr files. Does this mean people need to store both the netcdf AND the zarr files?

tennlee commented on packages/tutorial/src/pyearthtools/tutorial/HadisdDataClass.py:
```diff
@@ -0,0 +1,334 @@
+    #     return station_ids
+    
+
+    def get_all_station_ids(self, root_directory: Path | str = None) -> list[str]:
```
Here is says root_directory defaults to HADISD_HOME/netcdf, but in the 'filesystem' method it is called with a first argument of just HADISD_HOME, leading to inconsistencies.

tennlee commented on packages/tutorial/src/pyearthtools/tutorial/HadisdDataClass.py:
```diff
@@ -0,0 +1,334 @@
+
+        # Retrieve all station IDs from the dataset directory if "all" is present
+        if "all" in station_ids:
+            station_ids = self.get_all_station_ids(HADISD_HOME)
```
This is where the class calls get_all_station_ids without the '/netcdf' part, and it's looking through the netcdf files rather than the zarr files to determine available station numbers

tennlee commented on packages/tutorial/src/pyearthtools/tutorial/HadisdDataClass.py:
```diff
@@ -0,0 +1,334 @@
+            (50000, 79999, "WMO_050000-079999"),
+            (80000, 99999, "WMO_080000-099999"),
+            (100000, 149999, "WMO_100000-149999"),
+            (150000, 199999, "WMO_150000-199999"),
```
When unpacking the zarr files, why not just put everything in a single directory? Do the directories help particularly?

millerjoel replied:
Stephen made the same comment actually. There is no reason for this and your suggestion is sensible I think. I think it was purely done to break up the download into smaller chunks.

millerjoel committed:
Make downloads more robust and idempotent so repeat downloads don't occur

millerjoel committed:
Prevent conversion to zarr for already converted netcdf files

millerjoel committed:
Clear notebook output

tennlee commented:
Hi everyone. As discussed, I will shortly be merging this code and this will close the pull request. However, the various comments still need to be addressed. Joel has agreed to continue to work through the feedback, but the code is also useful enough to offer value to users as-is.

tennlee commented on packages/tutorial/src/pyearthtools/tutorial/__init__.py:
```diff
@@ -19,21 +19,24 @@
 """
 
-__version__ = "0.2.0"
+__version__ = "0.1.0"
```
```suggestion
__version__ = "0.2.0"
```

tennlee committed:
Update packages/tutorial/src/pyearthtools/tutorial/__init__.py
